File: mvvm_OpenGL/app/src/main/java/com/example/myopengl/MainActivity.java
package com.example.myopengl;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.hjq.permissions.OnPermission;
import com.hjq.permissions.Permission;
import com.hjq.permissions.XXPermissions;
import java.util.ArrayList;
import java.util.List;

public class MainActivity extends BaseActivity implements View.OnClickListener {

    private RecyclerView mList;
    private ArrayList<MenuBean> data;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        XXPermissions.with(this)
                .constantRequest()
                .permission(Permission.READ_EXTERNAL_STORAGE,Permission.WRITE_EXTERNAL_STORAGE,Permission.CAMERA)
                .request(new OnPermission() {
                    @Override public void hasPermission(List<String> granted, boolean all) {

                    }

                    @Override public void noPermission(List<String> denied, boolean quick) {

                    }
                });
        mList= (RecyclerView)findViewById(R.id.mList);
        mList.setLayoutManager(new LinearLayoutManager(this,LinearLayoutManager.VERTICAL,false));
        data=new ArrayList<>();
        add("绘制形体",FGLViewActivity.class);
        add("图片处理",SGLViewActivity.class);
        add("图形变换",VaryActivity.class);
        add("相机",CameraActivity.class);
        add("相机2 动画",Camera2Activity.class);
        add("相机3 美颜",Camera3Activity.class);
        add("压缩纹理动画",ZipActivity.class);
        add("FBO使用", FBOActivity.class);
        add("EGL后台处理",EGLBackEnvActivity.class);
        add("3D obj模型",ObjLoadActivity.class);
        add("obj+mtl模型",ObjLoadActivity2.class);
        add("VR效果",VrContextActivity.class);
        add("颜色混合",BlendActivity.class);
        add("光照",LightActivity.class);
        mList.setAdapter(new MenuAdapter());
    }

    private void add(String name, Class<?> clazz){
        MenuBean bean=new MenuBean();
        bean.name=name;
        bean.clazz=clazz;
        data.add(bean);
    }

    private class MenuBean{

        String name;
        Class<?> clazz;

    }

    private class MenuAdapter extends RecyclerView.Adapter<MenuAdapter.MenuHolder>{


        @Override
        public MenuHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            return new MenuHolder(getLayoutInflater().inflate(R.layout.item_button,parent,false));
        }

        @Override
        public void onBindViewHolder(MenuHolder holder, int position) {
            holder.setPosition(position);
        }

        @Override
        public int getItemCount() {
            return data.size();
        }

        class MenuHolder extends RecyclerView.ViewHolder{

            private Button mBtn;

            MenuHolder(View itemView) {
                super(itemView);
                mBtn= (Button)itemView.findViewById(R.id.mBtn);
                mBtn.setOnClickListener(MainActivity.this);
            }

            public void setPosition(int position){
                MenuBean bean=data.get(position);
                mBtn.setText(bean.name);
                mBtn.setTag(position);
            }
        }

    }

    @Override
    public void onClick(View view){
        int position= (int)view.getTag();
        MenuBean bean=data.get(position);
        startActivity(new Intent(this,bean.clazz));
    }
}

File: mvvm_OpenGL/app/src/main/java/com/example/mvvm_opengl/view/view/MainActivity.kt
package com.example.mvvm_opengl.view.view

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.databinding.ViewDataBinding
import com.aleyn.mvvm.base.BaseActivity
import com.aleyn.mvvm.base.NoViewModel
import com.blankj.utilcode.util.ToastUtils
import com.example.mvvm_opengl.R
import com.hjq.permissions.OnPermission
import com.hjq.permissions.Permission
import com.hjq.permissions.XXPermissions
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : BaseActivity<NoViewModel, ViewDataBinding>(), View.OnClickListener {
    override fun layoutId(): Int {
        return R.layout.activity_main
    }

    override fun initView(savedInstanceState: Bundle?) {
        btn_draw_triangle.setOnClickListener(this)
        btn_formal_draw_triangle.setOnClickListener(this)
    }

    override fun initData() {
        applyPermissions()
    }

    private fun applyPermissions() {
        XXPermissions.with(this)
            .constantRequest()
            .permission(
                Permission.READ_EXTERNAL_STORAGE,
                Permission.WRITE_EXTERNAL_STORAGE,
                Permission.CAMERA,
                Permission.RECORD_AUDIO)
            .request(object : OnPermission {
                override fun hasPermission(granted: List<String>, all: Boolean) {}
                override fun noPermission(denied: List<String>, quick: Boolean) {}
            })

    }

    override fun onClick(v: View?) {
        when(v?.id){
           R.id.btn_draw_triangle  ->{
               ToastUtils.showShort("常規三角形")
               startActivity(Intent(this,TriangleActivity::class.java))
           }
            R.id.btn_formal_draw_triangle ->{
                ToastUtils.showShort("正規三角形")
            }
        }
    }
}